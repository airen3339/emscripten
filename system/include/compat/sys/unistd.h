#include <unistd.h>
#include <random.h> // for getentropy()
