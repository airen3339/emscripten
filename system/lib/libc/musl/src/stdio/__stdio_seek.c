#include "stdio_impl.h"

off_t __stdio_seek(FILE *f, off_t off, int whence)
{
	off_t ret;
#ifdef SYS__llseek
	off_t off_hi =
#ifndef __EMSCRIPTEN__
		off >> 32;
#else
		0;
#endif
	if (syscall(SYS__llseek, f->fd, off_hi, off, &ret, whence)<0)
		ret = -1;
#else
	ret = syscall(SYS_lseek, f->fd, off, whence);
#endif
	return ret;
}
