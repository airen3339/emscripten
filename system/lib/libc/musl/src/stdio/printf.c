#include "stdio_impl.h"
#include <stdio.h>
#include <stdarg.h>

int printf(const char *restrict fmt, ...)
{
	int ret;
	va_list ap;
	va_start(ap, fmt);
	ret = vfprintf(stdout, fmt, ap);
	va_end(ap);
	return ret;
}

// XXX EMSCRIPTEN
int iprintf(const char *restrict fmt, ...)
{
	int ret;
	va_list ap;
	va_start(ap, fmt);
	ret = vfiprintf(stdout, fmt, ap);
	va_end(ap);
	return ret;
}

extern int __small_vfprintf(FILE *restrict f, const char *restrict fmt, va_list ap);

int __small_printf(const char *restrict fmt, ...)
{
	int ret;
	va_list ap;
	va_start(ap, fmt);
	ret = __small_vfprintf(stdout, fmt, ap);
	va_end(ap);
	return ret;
}

