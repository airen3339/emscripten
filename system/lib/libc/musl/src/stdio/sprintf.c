#include "stdio_impl.h"
#include <stdio.h>
#include <stdarg.h>

int sprintf(char *restrict s, const char *restrict fmt, ...)
{
	int ret;
	va_list ap;
	va_start(ap, fmt);
	ret = vsprintf(s, fmt, ap);
	va_end(ap);
	return ret;
}

// XXX EMSCRIPTEN
int siprintf(char *restrict s, const char *restrict fmt, ...)
{
	int ret;
	va_list ap;
	va_start(ap, fmt);
	ret = vsiprintf(s, fmt, ap);
	va_end(ap);
	return ret;
}

extern int __small_vsprintf(char *restrict s, const char *restrict fmt, va_list ap);

int __small_sprintf(char *restrict s, const char *restrict fmt, ...)
{
	int ret;
	va_list ap;
	va_start(ap, fmt);
	ret = __small_vsprintf(s, fmt, ap);
	va_end(ap);
	return ret;
}

