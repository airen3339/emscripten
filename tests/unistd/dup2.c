/*
 * Copyright 2021 The Emscripten Authors.  All rights reserved.
 * Emscripten is available under two separate licenses, the MIT license and the
 * University of Illinois/NCSA Open Source License.  Both these licenses can be
 * found in the LICENSE file.
 */

#include <assert.h>
#include <errno.h>
#include <fcntl.h>
#include <stdio.h>
#include <unistd.h>
#include <wasi/api.h>

#define __WASI_ERRNO_BADF (UINT16_C(8))

int main() {
  int f = 1;

  printf("DUP\n");
  int f2 = dup(f);

  // Print to stdout and print to fd = 3
  assert(f2 == 3);
  dprintf(f, "STDOUT\n");
  dprintf(f2, "CAN PRINT TO STDOUT WITH fd = 3\n");

  // Try calling dup with an invalid fd
  int f_invalid = dup(-1);
  assert(f_invalid == -1);

  printf("DUP2\n");
  int f3 = 5;
  int f4 = dup2(f, f3);

  // Print to stdout and print to fd = 4
  assert(f3 == 5);
  assert(f4 == 5);
  dprintf(f, "STDOUT\n");
  dprintf(f3, "CAN PRINT TO STDOUT WITH fd = 5\n");
  dprintf(f4, "CAN PRINT TO STDOUT WITH fd = 5\n");

  // Try printing to a non-existent fd
  errno = 0;
  dprintf(10, "FD DOES NOT EXIST\n");
  assert(errno == EBADF);

  // Try setting newfd == oldfd
  int f5 = dup2(f4, 5);
  assert(f5 == 5);
  dprintf(f5, "CAN PRINT TO STDOUT WITH fd = 5\n");

  // Try calling dup2 with an invalid newfd
  f5 = dup2(f4, -1);
  assert(f5 == -1);

  // Try calling dup2 with an invalid oldfd
  int f6 = dup2(-1, f5);
  assert(f6 == -1);

  // Try assigning a large fd
  int f7 = 4069;
  int f8 = dup2(f4, f7);

  dprintf(f8, "CAN PRINT TO STDOUT WITH f8 = 4069\n");

  int f9 = dup(-1);
  assert(f9 == -1);

  return 0;
}
