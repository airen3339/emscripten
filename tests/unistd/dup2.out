STDOUT
CAN PRINT TO STDOUT WITH fd = 3
DUP2
STDOUT
CAN PRINT TO STDOUT WITH fd = 5
CAN PRINT TO STDOUT WITH fd = 5
CAN PRINT TO STDOUT WITH fd = 5
CAN PRINT TO STDOUT WITH f8 = 4069
